### PR TITLE
Remove AutoService in favor of arcs_kt_particles

### PR DESCRIPTION
**Pull request by cromwellian**
Java annotation processors + bazel + kotlin + android are too much trouble.

* Remove @AutoService -- all it did for us is write a line in META-INF/services/className in a JAR
* Leverage existing arcs_kt_particles macro for isolatable particles. 
* Revert all hackery around Java 1.7 -> Java 1.8

Since this macro is used for WASM, by definition, it's only particles destined for DevHost or ProdHost, but not an external host. So now, by default, any particle compiled with arcs_kt_particles will run on ProdHost in Particle Accelerator. 

Note, you can still use @TargetHost to override this, but by default, auto-discoverable particles choose that as their home.

### CODE REVIEW

File: javatests/arcs/core/host/BUILD [hunk 1]
```diff
@@ -53,23 +54,47 @@ arcs_kt_plan(
...
 
-arcs_kt_jvm_library(
-    name = "particle",
-    testonly = 1,
-    srcs = glob(
-        ["*.kt"],
-        exclude = ["*Test.kt"],
-    ),
+arcs_kt_particles(
+    name = "pure-particles",
```
Comment (Cypher1) [inline javatests/arcs/core/host/BUILD]: Is this rename necessary? (i.e. I don't think it is but can't tell)

Comment (cromwellian) [reply javatests/arcs/core/host/BUILD]: Yes, because PurePerson depends on TestJvmProdHost, but the target that used to glob TestJvmProdHost also depends on PurePerson, so you have a cycle. I had to factor the host into its own target so PurePerson could target it. 

Alternatively, I could have made a custom annotation, but it makes sense to separate hosts and particles into different targets anyway.

File: third_party/java/arcs/build_defs/internal/kotlin_serviceloader_registry.bzl [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        output = ctx.outputs.out,
+        substitutions = {
+            "{particles}": ctx.attr.particles,
+        },
+    )
+
+kotlin_serviceloader_registry = rule(
+    implementation = _kotlin_serviceloader_registry,
+    attrs = {
+        "particles": attr.string(doc = "Newline separated list of fully qualified particle classes"),
```
Comment (csilvestrini) [inline third_party/java/arcs/build_defs/internal/kotlin_serviceloader_registry.bzl]: This would ideally be `attr.string_list` instead

File: third_party/java/arcs/build_defs/internal/kotlin_serviceloader_registry_template.tmpl [hunk 3]
```diff
@@ -0,0 +1,2 @@
+{particles}
```
Comment (csilvestrini) [inline third_party/java/arcs/build_defs/internal/kotlin_serviceloader_registry_template.tmpl]: If there's nothing else to add to this file, you can use `actions.write` instead of `actions.expand_template`.

Possibly there's a nice comment you'd want to add to this generated file, which might make it worth keeping the template.

Comment (cromwellian) [reply third_party/java/arcs/build_defs/internal/kotlin_serviceloader_registry_template.tmpl]: I didn't think the file format permitted comments, but apparently it does

https://docs.oracle.com/javase/tutorial/ext/basics/spi.html#packaging-the-dictionary-service-in-a-jar-file

Comment (cromwellian) [reply third_party/java/arcs/build_defs/internal/kotlin_serviceloader_registry_template.tmpl]: Changed to actions.write

File: third_party/java/arcs/build_defs/internal/kotlin_serviceloader_registry.bzl (hunk 2 above)
Comment (cromwellian) [reply third_party/java/arcs/build_defs/internal/kotlin_serviceloader_registry.bzl]: Done

File: third_party/java/arcs/build_defs/internal/kotlin.bzl [hunk 4]
```diff
@@ -177,42 +185,48 @@ def arcs_kt_library(
...
+            testonly = testonly,
             # Exclude any jvm-specific srcs.
             srcs = [src for src in srcs if not src.endswith(".jvm.kt")],
-            deps = [_to_wasm_dep(dep) for dep in deps],
-            testonly = testonly,
             visibility = visibility,
+            deps = [_to_wasm_dep(dep) for dep in deps],
         )
 
+def _extract_particle_name(src):
```
Comment (alxrsngrtn) [inline third_party/java/arcs/build_defs/internal/kotlin.bzl]: Let's put this at the bottom of the file w the other private utilities.
